Ray_Serve/CV-Serve.ipynb

```diff
@@ -0,0 +1,1173 @@
+{
+ "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "c399be64-44d3-45e4-9b43-3757a9b92daf",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "import ray\n",
+    "from ray.air.config import ScalingConfig\n",
+    "from ray import serve\n",
+    "import requests, json\n",
+    "from starlette.requests import Request\n",
+    "from typing import Dict\n",
+    "from transformers import SegformerForSemanticSegmentation, SegformerFeatureExtractor\n",
+    "from PIL import Image, ImageEnhance\n",
+    "import numpy as np\n",
+    "import matplotlib.pyplot as plt\n",
+    "import torch\n",
+    "import pickle\n",
+    "from io import BytesIO"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "7f391603-346b-4f08-a5b9-dee050b77409",
+   "metadata": {},
+   "source": [
+    "# Ray Serve\n",
+    "\n",
+    "## Intro\n",
+    "\n",
+    "### Outline\n",
+    "\n",
+    "-   Deployments\n",
+    "    -   Resources (CPU/GPU/custom)\n",
+    "    -   Runtime environments support, usage (functionality)\n",
+    "    -   Bound deployments, ServeHandles\n",
+    "-   Scaling and Performance\n",
+    "    -   Replicas\n",
+    "        -   num_replicas, autoscaling_config, max_concurrent_queries\n",
+    "    -   Request batching\n",
+    "-   Composition Patterns\n",
+    "    -   Imperative\n",
+    "    -   Declarative / Graph Deployment API\n",
+    "-   Architecture / Under-the-hood\n",
+    "    -   Ray cluster perspective - processes / workers / actors\n",
+    "    -   Request routing, queuing, load balancing in Serve\n",
+    "\n",
+    "### Example scenario: computer vision services\n",
+    "\n",
+    "For our example use case, we’ll see how to leverage Ray Serve to host a CV segmentation\n",
+    "model and how to enhance it using additional services such as image preprocessing.\n",
+    "\n",
+    "### Context: Ray AIR\n",
+    "\n",
+    "Ray AIR is the Ray AI Runtime, a set of high-level easy-to-use APIs for\n",
+    "ingesting data, training models – including reinforcement learning\n",
+    "models – tuning those models and then serving them.\n",
+    "\n",
+    "<img src=\"https://technical-training-assets.s3.us-west-2.amazonaws.com/Introduction_to_Ray_AIR/e2e_air.png\" width=600 loading=\"lazy\"/>\n",
+    "\n",
+    "Key principles behind Ray and Ray AIR are\n",
+    "* Performance\n",
+    "* Developer experience and simplicity"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "dfc41d8f-4332-4359-b957-f26bc79de7a3",
+   "metadata": {},
+   "source": [
+    "# Ray Serve\n",
+    "\n",
+    "Serve is a microservices framework for serving ML – the model serving\n",
+    "component of Ray AIR.\n",
+    "\n",
+    "<img src='https://technical-training-assets.s3.us-west-2.amazonaws.com/Ray_Serve/serve_architecture.png' width=700/>\n",
```

Comment:
This diagram makes it look like each node has n replicas of a deployment. It's more like n total replicas of the deployment are spread across m nodes. (where m could be >,=,< n)